awx/main/tasks/handlers/host_metric_summary_monthly_task.py

```diff
@@ -0,0 +1,169 @@
+    - license_consumed is license_consumed(previous month) + hosts_added - hosts_deleted
+    - - license_consumed for first updated month is computed also from all host metrics created before this month
+    - license_capacity is set only for current month, and it's never updated (value taken from current subscription)
+    """
+
+    def __init__(self):
+        self.host_metrics = {}
+        self.processed_month = self._get_first_month()
+        self.existing_summaries = None
+        self.existing_summaries_idx = 0
+        self.existing_summaries_cnt = 0
+        self.records_to_create = []
+        self.records_to_update = []
+
+    def execute(self):
+        self._load_existing_summaries()
+        self._load_hosts_added()
+        self._load_hosts_deleted()
+
+        # Get first month after last hard delete
+        month = self._get_first_month()
+        license_consumed = self._get_license_consumed_before(month)
+
+        # Fill record for each month
+        while month <= datetime.date.today().replace(day=1):
+            summary = self._find_or_create_summary(month)
+            # Update summary and update license_consumed by hosts added/removed this month
+            self._update_summary(summary, month, license_consumed)
+            license_consumed = summary.license_consumed
+
+            month = month + relativedelta(months=1)
+
+        # Create/Update stats
+        HostMetricSummaryMonthly.objects.bulk_create(self.records_to_create, batch_size=1000)
+        HostMetricSummaryMonthly.objects.bulk_update(self.records_to_update, ['license_consumed', 'hosts_added', 'hosts_deleted'], batch_size=1000)
+
+    def _get_license_consumed_before(self, month):
+        license_consumed = 0
+        for metric_month, metric in self.host_metrics.items():
+            if metric_month < month:
+                hosts_added = metric.get('hosts_added', 0)
+                hosts_deleted = metric.get('hosts_deleted', 0)
+                license_consumed = license_consumed + hosts_added - hosts_deleted
+            else:
+                break
+        return license_consumed
+
+    def _load_existing_summaries(self):
+        """Find all summaries newer than host metrics delete threshold"""
+        self.existing_summaries = HostMetricSummaryMonthly.objects.filter(date__gte=self._get_first_month()).order_by('date')
+        self.existing_summaries_idx = 0
+        self.existing_summaries_cnt = len(self.existing_summaries)
+
+    def _load_hosts_added(self):
+        """
+        SELECT 1 as id, date_trunc('month', first_automation) as month,
+               count(first_automation) AS hosts_added
+        FROM main_hostmetric
+        GROUP BY month
+        ORDER by month;
+        """
+        result = (
+            HostMetric.objects.annotate(month=TruncMonth('first_automation'))
+            .values('month')
+            .annotate(hosts_added=Count('first_automation'))
+            .values('month', 'hosts_added')
```

Comment:
@AlanCoding it doesn't produce the same sql, tests fail with this query

Reply:
You can get the query by:

```python
print(result.query)
```

This produces

```sql
SELECT DATE_TRUNC(MONTH, "main_hostmetric"."first_automation" AT TIME ZONE UTC) AS "month",
       COUNT("main_hostmetric"."first_automation") AS "hosts_added"
FROM "main_hostmetric"
GROUP BY "main_hostmetric"."id",
         1
ORDER BY 1 ASC
```

The difference that my modification introduces is `GROUP BY "main_hostmetric"."id", 1` as opposed to your query which has `GROUP BY 1` only. I do not know what high-level objective this affects.

Reply:
I don't know what to say - I printed it directly from the `awx-manage host_metric_summary_monthly` - if they look like your query, tests definitely won't work. The intention is to produce summaries by months

Reply:
also when I run it from `awx-manage shell_plus` it produces what I wrote

Reply:
`GROUP BY 1` is a valid SQL construction and is interchangeable with 
`GROUP BY DATE_TRUNC(MONTH, "main_hostmetric"."first_automation" AT TIME ZONE UTC)`.
It means "Group by 1st column". 

That's why it could produce different sqls for me and for you. And also it's an answer why it's important to keep it as it is.

Reply:
Yeah, the construction of the query is making sense to me now. Thanks. For the record, I still don't see why it would produce different sqls for me and you, unless you're running a different version of Django. Even if they are equivalent.

Reply:
good question. I'm running it in the docker container tools_awx_1, so there should be the Django from requirements.txt